### PR TITLE
Performance and reliability fixes for flow analysis

### PR DESCRIPTION
**Pull request by mavasani**
1. Combine Null and PointsToAnalysis into a single analysis. Null state for entities is now tracked as part of PointsTo analysis.
2. Fix the hang for the new DFA rules by being more conservating in tracking PointsTo values during merge.
3. Remove lot of unncessary null checks in Merge and Compare functions and add asserts to validate inputs and output of these functions.

### CODE REVIEW

File: src/Microsoft.CodeQuality.Analyzers.Exp/Core/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs [hunk 1]
```diff
@@ -98,37 +152,191 @@ protected override void SetValueForParameterOnExit(IParameterSymbol parameter, A
...
                 {
                     return PointsToAbstractValue.NoLocation;
                 }
 
                 var location = AbstractLocation.CreateAllocationLocation(operation, analysisEntity.Type);
-                return new PointsToAbstractValue(location);
+                return PointsToAbstractValue.Create(location, mayBeNull: true);
+            }
+
+            #region Predicate analysis
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers.Exp/Core/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs]: Code moved from `NullAnalysisOperationVisitor` to here.

File: src/Microsoft.CodeQuality.Analyzers.Exp/Core/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs [hunk 2]
```diff
@@ -342,19 +571,115 @@ private static PointsToAbstractValue VisitInvocationCommon(IOperation operation)
...
             }
 
             public override PointsToAbstractValue VisitDynamicInvocation(IDynamicInvocationOperation operation, object argument)
             {
                 var _ = base.VisitDynamicInvocation(operation, argument);
-                return VisitInvocationCommon(operation);
+                return VisitInvocationCommon(operation, operation.Operation);
+            }
+
+            private NullAbstractValue GetNullStateBasedOnInstanceOrReferenceValue(IOperation referenceOrInstance, ITypeSymbol operationType, NullAbstractValue defaultValue)
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers.Exp/Core/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs]: Code moved from `NullAnalysisOperationVisitor` to here.

File: src/Microsoft.CodeQuality.Analyzers.Exp/Core/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs [hunk 3]
```diff
@@ -98,37 +152,191 @@ protected override void SetValueForParameterOnExit(IParameterSymbol parameter, A
...
+
+                if (!equals)
+                {
+                    if (value != NullAbstractValue.Invalid && negatedValue != NullAbstractValue.Invalid)
+                    {
+                        var temp = value;
+                        value = negatedValue;
+                        negatedValue = temp;
+                    }
+                }
```
Comment (ivanbasov) [inline src/Microsoft.CodeQuality.Analyzers.Exp/Core/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs]: Could you please add some comments about the algorithm?

File: src/Microsoft.CodeQuality.Analyzers.Exp/Core/FlowAnalysis/Framework/DataFlow/AnalysisEntityMapAbstractDomain.cs [hunk 4]
```diff
@@ -48,31 +52,67 @@ TValue GetMergedValueForEntityPresentInOneMap(AnalysisEntity key, TValue value)
...
+
                     if (key1.InstanceLocation.Equals(key2.InstanceLocation))
                     {
                         resultMap[key1] = mergedValue;
                     }
                     else
                     {
+                        if (key1.SymbolOpt == null || key1.SymbolOpt != key2.SymbolOpt)
+                        {
+                            // PERF: Do not add a add a new key-value pair to the resultMap for unrelated entities or non-symbol based entities.
```
Comment (ivanbasov) [inline src/Microsoft.CodeQuality.Analyzers.Exp/Core/FlowAnalysis/Framework/DataFlow/AnalysisEntityMapAbstractDomain.cs]: >add a add a [](start = 44, length = 11)

reduce

**Pull request merged**